xdsl/transforms/stencil_bufferize.py

```diff
@@ -0,0 +1,435 @@
+        uses = handle.uses.copy()
+        if len(uses) > 1:
+            return
+        user = uses.pop().operation
+
+        effecting = [
+            o
+            for o in walk_from_to(load, user)
+            if underlying in o.operands
+            and (not is_side_effect_free(o))
+            and (o not in (load, op, user))
+        ]
+        if effecting:
+            return
+
+        rewriter.replace_matched_op(new_ops=[], new_results=[underlying])
+
+
+class ApplyLoadStoreFoldPattern(RewritePattern):
+    """
+    If an apply is acting on a buffer, that is *only* used to be copied in another
+    buffer, just make the apply work on the latter directly.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: StoreOp, rewriter: PatternRewriter):
+        temp = op.temp
+
+        # We are looking for a loaded destination of an apply
+        if not isinstance(load := temp.owner, LoadOp):
+            return
+
+        infield = load.field
+
+        other_uses = [u for u in infield.uses if u.operation is not load]
+
+        if len(other_uses) != 1:
+            return
+
+        other_use = other_uses.pop()
+
+        if not isinstance(
+            apply := other_use.operation, ApplyOp
+        ) or other_use.index < len(apply.args):
+            print(other_use)
+            print()
+            return
+
+        # Get first occurence of the field, to walk from it
+        start = op.field.owner
+        if isinstance(start, Block):
+            if start is not op.parent:
+                return
+            start = cast(Operation, start.first_op)
+        effecting = [
+            o
+            for o in walk_from_to(start, op)
+            if infield in o.operands
+            and (not is_side_effect_free(o))
+            and (o not in (load, apply))
+        ]
+        if effecting:
+            print("effecting: ", effecting)
+            print(load)
+            return
+
+        new_operands = list(apply.operands)
+        new_operands[other_use.index] = op.field
+
+        new_apply = ApplyOp.create(
+            operands=new_operands,
+            result_types=[],
+            properties=apply.properties.copy(),
+            attributes=apply.attributes.copy(),
+            regions=[
+                Region(Block(arg_types=[SSAValue.get(a).type for a in apply.args])),
+            ],
+        )
+
+        rewriter.inline_block(
+            apply.region.block,
+            InsertPoint.at_start(new_apply.region.block),
+            new_apply.region.block.args,
+        )
+
+        new_load = LoadOp.create(
+            operands=[op.field],
+            result_types=[r.type for r in load.results],
+            attributes=load.attributes.copy(),
+            properties=load.properties.copy(),
+        )
+
+        rewriter.replace_op(apply, new_apply)
+        rewriter.replace_op(load, new_load)
+        rewriter.erase_op(op)
```

Comment:
Question to my rested self

Why is the Apply considered here already? Can't it just be a LoadStoreFold ?

Reply:
There are ways but it requires finer analysis, on shapes as on more precise side effects.
So, ditching for now, as the first goal is only to get it satisfyingly working on existing stencil workloads.